### DISCUSSION TITLE
How do you deal with deploying mid-job?

### DISCUSSION THREAD

**Discussion by charliemday** [django-q2/django-q2#178] [Q&A]
I have a Django Background job running on a server where the jobs can take anywhere between 10 - 60 mins. I want to be able to deploy without interrupting the job (and it's just left hanging). How have people got around this?

**Comment by GDay** [comment]
I don't think there is a good way to get around this, except for killing the process manually.

I generally prefer to break jobs up in smaller jobs. For example: when deleting objects that have files attached. I create a task for each object, so they only take a few seconds each, instead of multiple minutes for all. This also means that I can do deploys in between. But I also realize that that might not work for every job.

The jobs you are suggesting, are they safe to be cancelled while they are running? 
If so, an argument could be added to the `manage.py qcluster` function to just kill the worker processes regardless if it's running something (when qcluster gets stopped).

I don't think there is an easy way around this right now.

**Reply by charliemday** [reply]
Right now the jobs could take upwards of 1 hour and they can't be cancelled safely (i.e. they'd lose their state). For example, A single job makes ~10 requests to an API every second asynchronously and this could be upwards of ~25K requests so could be quite tricky to split this up (and the order has to be respected and there are calls to the database).

I was wondering if there was a way to basically have 2 servers for background jobs - essentially a **primary** and **secondary** server.  

The **primary** server handles all main requests until a deployment is triggered. When a deployment is triggered the **primary** server becomes the **secondary** server (which only completes all in progress jobs and doesn't accept any new jobs) and the original **secondary** server becomes the new **primary** server with the latest deployed code and accepts all new incoming job requests.

Does this approach make sense (I'd love a second opinion in case I'm missing something stupid)?

**Reply by GDay** [reply]
That makes sense, but I think this is a task for your deployment workflow and not for Django-q(2). For example, I think something like this covers your use case: https://dokku.com/docs/deployment/zero-downtime-deploys/. I don't know what you use for deployment, but I think most providers have some kind of "zero downtime" deploys that work in a similar way. It will basically send the "stop" signal to the running container and spin up a new container (which will then take on the new tasks). When django-q2 receives the stop command it will not pickup any new tasks anymore.

The only thing to worry about here is data(base) migrations. If your worker is putting data in the database, while the database table got changed due to a new deploy, then the job could fail. But I don't think there is an easy solution for this. Potentially workable solution would be to stall the deployment until django-q2 has fully stopped. This would likely as well be a job for the deployment workflow (sending the stop command to the container/server and then waiting till the process has ended).
